src/backend/distributed/executor/local_executor.c

```diff
@@ -468,48 +468,45 @@ ExtractLocalAndRemoteTasks(bool readOnly, List *taskList, List **localTaskList,
 		/* either the local or the remote should be non-nil */
 		Assert(!(localTaskPlacementList == NIL && remoteTaskPlacementList == NIL));
 
-		if (list_length(task->taskPlacementList) == 1)
+		if (list_length(localTaskPlacementList) > 0)
 		{
-			/*
-			 * At this point, the task has a single placement (e.g,. anchor shard
-			 * is distributed table's shard). So, it is either added to local or
-			 * remote taskList.
-			 */
-			if (localTaskPlacementList == NIL)
+			if (list_length(task->taskPlacementList) == 1)
 			{
-				*remoteTaskList = lappend(*remoteTaskList, task);
+				*localTaskList = lappend(*localTaskList, task);
 			}
 			else
 			{
-				*localTaskList = lappend(*localTaskList, task);
+				/*
+				 * At this point, we're dealing with reference tables or intermediate
+				 * results where the task has placements on both local and remote
+				 * nodes. We always prefer to use local placement, and require remote
+				 * placements only for modifications.
```

Comment:
> we're dealing with reference tables or intermediate results

we should probably remove this sentence, since there are other cases so it might be misleading.